src/server/session.ts

```diff
@@ -2015,4 +2018,40 @@ namespace ts.server {
             }
         }
     }
+
+    /* @internal */ // exported for tests only
+    /** Since we will perform the rename in a changed document, line and offset information must be adjusted if the new text adds/deletes newlines. */
+    export function fixupLineAndOffset(oldPosition: protocol.Location, oldText: string, location: number, locationFileName: string, edits: ReadonlyArray<FileTextChanges>): protocol.Location {
+        let { line, offset } = oldPosition;
+        const lineStartOfLocation = location - (offset - 1); // offset is 1-based, location is 0-based

```

Comment:
Will this value be incorrect if there's an edit on the same line, before `location`?

Reply:
This is meant to be the line start in the *old* document.

Reply:
But isn't `location` in the new document?